[native] Add remote function server end to end tests

pedroerp opened the pull request:
Because the new test suite requires a remote function server binary to
be available, it's only enabled when `-Ppresto-native-execution-remote-functions`
is passed to maven. When it's enabled, the binary path is configured in
the pom.xml file, or can be overwritten using the REMOTE_FUNCTION_SERVER
env variable.


```
== NO RELEASE NOTE ==
```

pedroerp commented on .circleci/continue_config.yml:
```diff
@@ -115,11 +115,12 @@ jobs:
           root: presto-native-execution
           paths:
             - _build/debug/presto_cpp/main/presto_server
+            - _build/debug/velox/velox/functions/remote/server/velox_functions_remote_server_main
```
Disregard these test changes for now (details on https://github.com/facebookincubator/velox/discussions/6163). Will undo them before merging.

mbasmanova commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeRemoteFunctions.java:
```diff
@@ -0,0 +1,73 @@
+    {
+        // Initialize the remote function thrift server process.
+        remoteFunctionServerBinaryPath = System.getProperty("REMOTE_FUNCTION_SERVER");
+        if (remoteFunctionServerBinaryPath != null) {
```
Why would we want to skip the test? Can this mask issues? Let's require that this property is set and fail the test it isn't.

pedroerp replied:
@mbasmanova happy to make the change, but the tricky part here is that the remote function binary not be available, for example, if the PRESTO_ENABLE_REMOTE_FUNCTIONS build flag was not specified. In that case, just running all unit tests would result in failures.

vermapratyush replied:
I suggest we introduce a new maven profile (using `maven-surefire-plugin` groups) for remote function tests. We do this for specialised tests like hive, parquet, and other meta internal integrations.

If we want to disable remote function tests by default, we could add:
`<excludeGroups>remote-functions</exludeGroups>`
to 
https://github.com/prestodb/presto/blob/6465ee8a6b57e653606dca6ca433db54e3f1dc53/presto-native-execution/pom.xml#L191


And introduce a new test suite exclusively for remote functions, like this:
```
          <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <configuration>
                    <forkCount>1</forkCount>
                    <reuseForks>false</reuseForks>
                    <groups>remote-functions</groups>
                    <systemPropertyVariables>
                        <PRESTO_SERVER>/root/project/build/debug/presto_cpp/main/presto_server</PRESTO_SERVER>
                        <DATA_DIR>/tmp/velox</DATA_DIR>
                    </systemPropertyVariables>
                </configuration>
            </plugin>
```

With this change we could then mark `REMOTE_FUNCTION_SERVER` as a require property in `preInitQueryRunners`.

pedroerp replied:
Nice! Let me look into this.

vermapratyush replied:
Something like this: https://github.com/prestodb/presto/pull/20658/commits/6868a3364589a8f38b8cd18dd0c0e02a6d78afda

pedroerp replied:
thanks @vermapratyush. I incorporated your changes. Could you take one last look?

mbasmanova commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeRemoteFunctions.java:
```diff
@@ -0,0 +1,73 @@
+public abstract class AbstractTestNativeRemoteFunctions
+        extends AbstractTestQueryFramework
+{
+    public String remoteFunctionServerUds;
```
Can these be private?

pedroerp replied:
I think so if I add a getter method. Let me check

mbasmanova commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeRemoteFunctions.java:
```diff
@@ -0,0 +1,73 @@
+
+        assertQuery("SELECT remote.schema.ceil(ceil(discount)) FROM lineitem", "SELECT ceil(ceil(discount * 7)) FROM lineitem");
+
+        // TODO: `throwOnError` needs to be implemented in the server so we can have remote functions as function parameters.
```
What happens now? Maybe create a GitHub issue and include the current error in the description, then add a link to the issue here?

pedroerp replied:
Today we get this exception:

https://github.com/facebookincubator/velox/blob/main/velox/functions/remote/server/RemoteFunctionService.cpp#L79

I have this listed as one item in this Issue:

https://github.com/facebookincubator/velox/issues/5288

mbasmanova commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/PrestoNativeQueryRunnerUtils.java:
```diff
@@ -40,6 +40,10 @@
 public class PrestoNativeQueryRunnerUtils
 {
     private static final String DEFAULT_STORAGE_FORMAT = "DWRF";
+    public static final String REMOTE_FUNCTION_UDS = "remote_function_server.socket";
```
What does UDS stand for?

pedroerp replied:
Unix Domain Socket

mbasmanova commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/PrestoNativeQueryRunnerUtils.java:
```diff
@@ -225,7 +250,30 @@ public static QueryRunner createNativeQueryRunner(boolean useThrift, String stor
+
+    public static String startRemoteFunctionServer(String remoteFunctionServerBinaryPath)
+    {
+        // Start the remote function server. Return the UDS path used to communicate with it.
```
Perhaps, move this into method-level comment.

mbasmanova commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/PrestoNativeQueryRunnerUtils.java:
```diff
@@ -225,7 +250,30 @@ public static QueryRunner createNativeQueryRunner(boolean useThrift, String stor
+            Path tempDirectoryPath = Files.createTempDirectory("RemoteFunctionServer");
+            Path remoteFunctionServerUdsPath = tempDirectoryPath.resolve(REMOTE_FUNCTION_UDS);
+
+            Logger log = Logger.get(PrestoNativeQueryRunnerUtils.class);
```
usually, this is defined as private static final member variable of the class.

mbasmanova reviewed (commented):
@pedroerp This looks great. Some comments below. What would it take to enable this test for Presto-on-Spark?

CC: @vermapratyush

pedroerp commented:
Thank you guys for the review. I'll take a look at the comments later today.

> What would it take to enable this test for Presto-on-Spark?

I think it should work by just adding the corresponding unit test file. Let me ensure this one works and land it first, then I can enable it in Presto-on-Spark on a follow up PR.

pedroerp commented:
@mbasmanova @vermapratyush thanks for the review. I addressed your comments, please take another look when you have a moment. 

There are two remaining points to be addressed still:

1. What to do when REMOTE_FUNCTION_SERVER is not set? 

The problem with failing invariably is that the remote function server binary itself will only be built if the user provides a "-DPRESTO_ENABLE_REMOTE_FUNCTIONS=ON" flag, which is OFF by default. So in the common case (when the flag OFF) the test would the fail. The current code skips the test if this flag is not set, but I'm open to suggestions.

2. Thrift symbol crash.

Because of https://github.com/facebookincubator/velox/discussions/6163 the CI tests crash when PRESTO_ENABLE_PARQUET is ON. We can either wait until we move thrift to fbthrift (like proposed in this discussion), or create an entire new job to compile only remote function, but not parquet, and run the tests. Thoughts?

vermapratyush commented:
@pedroerp Thanks for the change, this looks good to me. Minor cleanup after the changes made.

pedroerp commented:
@vermapratyush I'm still struggling with this. It seems like when I compile the code using the configuration you suggested, there's is something that does not get initialized and `getSession()` fails. I poked around for a while but couldn't find the reason; any ideas? This is the error I get:

```
[ERROR]   TestPrestoNativeRemoteFunctions>AbstractTestNativeRemoteFunctions.testRemoteFunction:66->AbstractTestQueryFramework.assertQuery:163->AbstractTestQueryFramework.getSession:123 » Null Pointer
```

Moreover, like you pointed out, the tests take a lot longer to run since they now load all other test suites, and somehow the output of the tests are not print to stdout anymore. 

The only thing I can think of at this point is to go back at the initial mode, which has the following semantic:

1. Is user specified the REMOTE_FUNCTION_SERVER binary, run the tests
2. If not, make the test throw the Skip exception. 

The problem with this approach being that if the user does not specify the env variable, the test will not run and might not signal a legit failure. However, any breakages would get caught in CI since we set the correct flags there. 

Thoughts?

vermapratyush reviewed (approved):
Thanks a lot @pedroerp for making the changes. The PR looks good to me.

Before landing, remember to remove the line `-Ppresto-native-execution-remote-functions`  from `.circleci/continue_config.yml`

mbasmanova reviewed (approved):
Thanks.
